core/ir/riscv64/isl/README.md

```diff
@@ -0,0 +1,143 @@
+# Instruction Set Listing
+
+## Format
+
+```text
+<name> | <instruction-format> | [field] | <fixedbits>
+```
+
+The file name will become the extension.
+There is a convention to specify fields in the reverse order of the assembly.
+This way the instruction macros are generated properly.
+
+If you want to add a new extension, add a new value to the `riscv64_isa_ext_t`
+enum in `codec.h`.
+
+When adding a new instruction make sure to update the `get_rvc_instr_info()` function in `codec.c` to properly decode the opcode.
+
+
+## Example
+
+```text
+add    | r  | rs2 rs1 rd    | 0000000..........000.....0110011
+c.lwsp | ci | clwsp_imm crd | 010...........10
+```
+
+
+## Instruction Format
+
+The instruction format should be one of the following:
+
+- r
+- r4
+- i
+- s
+- b
+- u
+- j
+- cr
+- ci
+- css
+- ciw
+- cl
+- cs
+- ca
+- cb
+- cj
+
+This maps to `riscv64_inst_fmt_t` enum in `codec.h` and `Format` enum in
+`codec.py` generator. However at this point the format field is not put into
+generated `instr_info_t` structures. This may change in the future, so in case
+of a new format, keep the above definitions in sync.
+
+
+## Field
+
+The list of available fields are written below.
+
+For non-compressed instructions:
+
+- rd
+- rd(fp)
+- rs1
+- rs1(fp)
+- base
+- rs2
+- rs2(fp)
+- rs3
+- fm
+- pred
+- succ
+- aqrl
+- csr
+- rm
+- shamt
+- shamt5
+- shamt6
+- csr_imm
+- i_imm
+- s_imm
+- b_imm
+- u_imm
+- j_imm
+- imm
+
+For compressed instructions:
+
+- crd
+- crd(fp)
+- crd_
+- crd_(fp)
+- crd__
+- crs1
+- crs1_
+- crs2
+- crs2(fp)
+- crs2_
+- crs2_(fp)
+- cshamt
+- caddi16sp_imm
+- clwsp_imm
+- cldsp_imm
+- clui_imm
+- cswsp_imm
+- csdsp_imm
+- ciw_imm
+- clw_imm
+- cld_imm
+- csw_imm
+- csd_imm
+- cimm5
+- cb_imm
+- cj_imm
+
+This maps into `riscv64_fld_t` enum in `codec.h` and `Field` enum in `codec.py`
+generator.
+
+If you want to add a new field:
+
+1. In `codec.py`:
+
+    1. Add new value to the `Field` enum.
+    2. If this is a destination field, update the `Field.is_dest()` function.
+    3. Update stringification functions: `Field.arg_name()` for C-macro argument
+       name and `Field.asm_name()` for assembly syntax comment override (if
+       necessary).
+    4. Add a new entry to the `cmt_str` list in `Field.arg_comment()` used as a
```

Comment:
Having cmt_str and sz_str separate from the fields is a bit confusing and error-prone. Can we combine these three somehow?

Reply:
Good idea. I'll combine those.

Reply:
Done. I've bundled all this info into Field object attributes. This makes it really hard to not fill all the required attributes when adding a new field.